### PR TITLE
Torchscriptable RNNT Loss

### PR DESCRIPTION
**Pull request by carolineechen**
- move RNNT loss autograd function to C++
- add torchscriptability tests

cc #1240

### CODE REVIEW

File: torchaudio/csrc/rnnt/autograd.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      tgt_lengths,
+      blank,
+      clamp,
+      fused_log_smax,
+      reuse_logits_for_grads);
+  return std::make_tuple(results[0], results[1]);
+}
+
+TORCH_LIBRARY_IMPL(torchaudio, Autograd, m) {
+  m.impl("rnnt_loss", rnnt_loss_autograd);
```
Comment (vincentqb) [inline torchaudio/csrc/rnnt/autograd.cpp:71]: does binding RNNTLossFunction::apply work directly?

Comment (carolineechen) [reply torchaudio/csrc/rnnt/autograd.cpp:70]: It doesn't immediately work because of some C++ errors and I think the return types would need to be changed also

File: torchaudio/prototype/rnnt_loss.py [hunk 2]
```diff
@@ -166,17 +105,30 @@ def rnnt_loss(
...
-        blank,
-        clamp,
-        fused_log_softmax,
-        reuse_logits_for_grads,
-    )
-    return cost
+    # move everything to the same device.
+    targets = targets.to(device=logits.device)
+    logit_lengths = logit_lengths.to(device=logits.device)
+    target_lengths = target_lengths.to(device=logits.device)
```
Comment (vincentqb) [inline torchaudio/prototype/rnnt_loss.py:111]: i can see a situation where a user moved all but the logits accidentally, and would encounter a very slow computation of the loss function. do we do this elsewhere in the codebase? were the checks in the c++ code not catching this already and raising an error?

File: torchaudio/prototype/rnnt_loss.py [hunk 3]
```diff
@@ -166,17 +105,30 @@ def rnnt_loss(
...
+
+    # make sure all int tensors are of type int32.
+    targets = targets.int()
+    logit_lengths = logit_lengths.int()
+    target_lengths = target_lengths.int()
```
[the remark this answers is not in the document]
Comment (carolineechen) [reply torchaudio/prototype/rnnt_loss.py]: you're right, the C++ checks should catch these failures

**Review by mthrok — APPROVED** [review]
Looks good.

File: torchaudio/prototype/rnnt_loss.py [hunk 4]
```diff
@@ -166,17 +87,20 @@ def rnnt_loss(
...
-        logit_lengths,
-        target_lengths,
-        blank,
-        clamp,
-        fused_log_softmax,
-        reuse_logits_for_grads,
-    )
-    return cost
+    if blank < 0:  # reinterpret blank index if blank < 0.
+        blank = logits.shape[-1] + blank
```
Comment (vincentqb) [inline torchaudio/prototype/rnnt_loss.py:91]: is this also needed for `alphas` and `betas`?

Comment (carolineechen) [reply torchaudio/prototype/rnnt_loss.py:91]: It seems likely that this may be the case, but will leave this as a followup. Per offline discussion, it may be a better idea to just remove these functions which are not used internally and also not exposed to the user.
